src/net/ossl.cc

```diff
@@ -215,9 +217,59 @@ using pkcs12 = ssl_handle<PKCS12, PKCS12_free>;
 using ssl_ctx_ptr = ssl_handle<SSL_CTX, SSL_CTX_free>;
 using ssl_ptr = ssl_handle<SSL, SSL_free>;
 using x509_verify_param_ptr = ssl_handle<X509_VERIFY_PARAM, X509_VERIFY_PARAM_free>;
+using ssl_session_ptr = ssl_handle<SSL_SESSION, SSL_SESSION_free>;
+
+/**
+ * The purpose of this structure is to hold the AES and HMAC keys used to encrypt/decrypt TLS 1.3
+ * session tickets given to and presented by, the TLS client.
+ */
+struct session_ticket_keys {
+    std::array<uint8_t, 32> aes_key;
+    std::array<uint8_t, 32> hmac_key;
+
+    void generate_keys() {
+        generate_key(aes_key);
+        generate_key(hmac_key);
+    }
+
+    ~session_ticket_keys() {
+        clear_key(aes_key);
+        clear_key(hmac_key);
+    }
+
+private:
+    /**
+     * This will zeroize the key contents by writing 0xff, then 0x00,
+     * and finally 0xff over the memory that held the key
+     * @tparam N Size of the array
+     * @param key The key to clear
+     */
+    template<std::size_t N>
+    static void clear_key(std::array<uint8_t, N>& key) {
+        // Zeroize sensitive key data
+        memset(key.data(), 0xff, N);
+        memset(key.data(), 0x00, N);
+        memset(key.data(), 0xff, N);
+    }
+
```

Comment:
I was reviewing the PR of the GnuTLS implementation and there's a comment in the gnutls docs that the session ticket keys should be overwritten before calling free on them to avoid leaking the key. This seems to be missing from the gnutls implementation. Should we fix/report that? https://man7.org/linux/man-pages/man3/gnutls_session_ticket_enable_server.3.html

```
diff --git a/src/net/tls.cc b/src/net/tls.cc
index 90df569e..d8b101f0 100644
--- a/src/net/tls.cc
+++ b/src/net/tls.cc
@@ -299,6 +299,7 @@ struct gnutls_datum : public gnutls_datum_t {
     }
     ~gnutls_datum() {
         if (data != nullptr) {
+            ::gnutls_memset(data, 0, size);
             ::gnutls_free(data);
         }
     }
```

Reply:
nice catch there... yeah probably worthwhile to upstream a one liner for them

Reply:
Created https://github.com/scylladb/seastar/pull/2424 to fix this